### PR TITLE
Get rid of empty string of redirect type from database structure

### PR DESCRIPTION
**Pull request by zuk3975**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Related to PR - https://github.com/PrestaShop/PrestaShop/pull/20473. We did get rid of empty string as redirect_type in code, this PR takes care of sql upgrade and sql structure. We didn't discuss it yet, but it's also an option to leave it unchanged to avoid BC breaks and maybe remove it in some later versions :man_shrugging: 
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | part of #14776
| How to test?  | Needs to perform version upgrade from to from 1.7.7 to 1.7.8

**BC BREAKS**
removes EVAL `''` from product.redirect_type

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/22378)

### CODE REVIEW

**Review by Progi1984 — APPROVED** [review]

**Review by Progi1984 — CHANGES_REQUESTED** [review]
Travis is completely broken.

File: install-dev/data/db_structure.sql [hunk 1]
```diff
@@ -1647,7 +1647,7 @@ CREATE TABLE `PREFIX_product` (
   `text_fields` tinyint(4) NOT NULL DEFAULT '0',
   `active` tinyint(1) unsigned NOT NULL DEFAULT '0',
   `redirect_type` ENUM(
-    '', '404', '301-product', '302-product',
+    '404', '301-product', '302-product',
     '301-category', '302-category'
   ) NOT NULL DEFAULT '',
```
Comment (PierreRambaud) [inline install-dev/data/db_structure.sql]: You need to change the DEFAULT here too

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
The update script is not the same as the db_structure

**Review by Progi1984 — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

**Comment by SD1982** [comment]
Hi @zuk3975 I have thiis error during the upgrade process

`[INTERNAL] /var/www/html/ps177PR2/app/AppKernel.php line 39 - ParseError: syntax error, unexpected ';' #0 /var/www/html/ps177PR2/modules/autoupgrade/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/var/www/html/p...') #1 [internal function]: Composer\Autoload\ClassLoader->loadClass('AppKernel') #2 /var/www/html/ps177PR2/config/autoload.php(28): spl_autoload_call('AppKernel') #3 /var/www/html/ps177PR2/modules/autoupgrade/classes/UpgradeContainer.php(545): require_once('/var/www/html/p...') #4 /var/www/html/ps177PR2/modules/autoupgrade/classes/TaskRunner/Upgrade/UpgradeDb.php(73): PrestaShop\Module\AutoUpgrade\UpgradeContainer->initPrestaShopAutoloader() #5 /var/www/html/ps177PR2/admin-dev/autoupgrade/ajax-upgradetab.php(52): PrestaShop\Module\AutoUpgrade\TaskRunner\Upgrade\UpgradeDb->init() #6 {main}`

It's probably related to the confliciting file can you check to fix the confilct please ?

**Comment by PierreRambaud** [comment]
> Hi @zuk3975 I have thiis error during the upgrade process
> 
> `[INTERNAL] /var/www/html/ps177PR2/app/AppKernel.php line 39 - ParseError: syntax error, unexpected ';' #0 /var/www/html/ps177PR2/modules/autoupgrade/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/var/www/html/p...') #1 [internal function]: Composer\Autoload\ClassLoader->loadClass('AppKernel') #2 /var/www/html/ps177PR2/config/autoload.php(28): spl_autoload_call('AppKernel') #3 /var/www/html/ps177PR2/modules/autoupgrade/classes/UpgradeContainer.php(545): require_once('/var/www/html/p...') #4 /var/www/html/ps177PR2/modules/autoupgrade/classes/TaskRunner/Upgrade/UpgradeDb.php(73): PrestaShop\Module\AutoUpgrade\UpgradeContainer->initPrestaShopAutoloader() #5 /var/www/html/ps177PR2/admin-dev/autoupgrade/ajax-upgradetab.php(52): PrestaShop\Module\AutoUpgrade\TaskRunner\Upgrade\UpgradeDb->init() #6 {main}`
> 
> It's probably related to the confliciting file can you check to fix the confilct please ?

I confirm it's related to the conflict :sweat:

**Review by PierreRambaud — APPROVED** [review]

**Review by atomiix — APPROVED** [review]

**Comment by SD1982** [comment]
Hi @zuk3975 I still have an error during the upgrade process
`[INTERNAL] /var/www/html/ps177/app/AppKernel.php line 39 - ParseError: syntax error, unexpected ';' #0 /var/www/html/ps177/modules/autoupgrade/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/var/www/html/p...') #1 [internal function]: Composer\Autoload\ClassLoader->loadClass('AppKernel') #2 /var/www/html/ps177/config/autoload.php(28): spl_autoload_call('AppKernel') #3 /var/www/html/ps177/modules/autoupgrade/classes/UpgradeContainer.php(545): require_once('/var/www/html/p...') #4 /var/www/html/ps177/modules/autoupgrade/classes/TaskRunner/Upgrade/UpgradeDb.php(73): PrestaShop\Module\AutoUpgrade\UpgradeContainer->initPrestaShopAutoloader() #5 /var/www/html/ps177/admin-dev/autoupgrade/ajax-upgradetab.php(52): PrestaShop\Module\AutoUpgrade\TaskRunner\Upgrade\UpgradeDb->init() #6 {main}`

image

**Comment by PierreRambaud** [comment]
> Hi @zuk3975 I still have an error during the upgrade process
> `[INTERNAL] /var/www/html/ps177/app/AppKernel.php line 39 - ParseError: syntax error, unexpected ';' #0 /var/www/html/ps177/modules/autoupgrade/vendor/composer/ClassLoader.php(322): Composer\Autoload\includeFile('/var/www/html/p...') #1 [internal function]: Composer\Autoload\ClassLoader->loadClass('AppKernel') #2 /var/www/html/ps177/config/autoload.php(28): spl_autoload_call('AppKernel') #3 /var/www/html/ps177/modules/autoupgrade/classes/UpgradeContainer.php(545): require_once('/var/www/html/p...') #4 /var/www/html/ps177/modules/autoupgrade/classes/TaskRunner/Upgrade/UpgradeDb.php(73): PrestaShop\Module\AutoUpgrade\UpgradeContainer->initPrestaShopAutoloader() #5 /var/www/html/ps177/admin-dev/autoupgrade/ajax-upgradetab.php(52): PrestaShop\Module\AutoUpgrade\TaskRunner\Upgrade\UpgradeDb->init() #6 {main}`
> 
> image

Did you fetch the right branch without conflicts? Because I don't see how it could be related to this PR :sweat_smile:

**Comment by SD1982** [comment]
Yes I think I tested with the right branch but if in doubt I will redo the whole procedure from the beginning just to be sure

**Comment by SD1982** [comment]
I restarted the process from the beginning and I still have the same error
Step followed:
- install the PR on develop
- create a build
- install 1.7.7.1 branch
- try to upgrade

**Comment by zuk3975** [comment]
I've successfully upgraded using CLI script ( `install-dev/upgrade/upgrade.php`). 
I don't know how to provoke 1click module upgrade from UI :man_shrugging: 
Screenshot from 2021-01-06 14-56-02


**EDIT**
I've finally managed to upgrade it using UI 1click upgrade module and it succeeded. Also previous `''` values where changed to `404` as expected in database.
Screenshot from 2021-01-06 18-10-35

**Comment by matks** [comment]
> I've successfully upgraded using CLI script ( `install-dev/upgrade/upgrade.php`).
> I don't know how to provoke 1click module upgrade from UI 🤷‍♂️
> Screenshot from 2021-01-06 14-56-02

See https://devdocs.prestashop.com/1.7/basics/keeping-up-to-date/use-autoupgrade-module/ 😄

**Review by jolelievre — APPROVED** [review]
I think this BC break is acceptable

**Review by Progi1984 — APPROVED** [review]

**Comment by SD1982** [comment]
Hi @zuk3975 I don't know why but I still have the error during the upgrade procedure so it's @khouloudbelguith  who performed the PR test and everything seems ok
Thank you @zuk3975 !!

**Comment by PierreRambaud** [comment]
Thank you @zuk3975

**Pull request merged**